src/main/java/com/algorand/algosdk/transaction/AtomicTransactionComposer.java

```diff
@@ -0,0 +1,439 @@
+            if (!resp.isSuccessful()) {
+                retList.add(new ReturnValue(
+                        null,
+                        null,
+                        null,
+                        resp.message()
+                ));
+                continue;
+            }
+
+            PendingTransactionResponse respBody = resp.body();
+            List<byte[]> logs = respBody.logs;
+            byte[] retLine = null;
+            for (int logIndex = logs.size() - 1; logIndex >= 0; logIndex--) {
+                byte[] line = logs.get(logIndex);
+                if (!checkLogRet(line))
+                    continue;
+                retLine = line;
+                break;
+            }
+            if (retLine == null) {
+                retList.add(new ReturnValue(
+                        null,
+                        null,
+                        null,
+                        "expected to find logged return value, got none"
+                ));
+                continue;
+            }
+
+            byte[] abiEncoded = Arrays.copyOfRange(retLine, 4, retLine.length);
+            Object decoded = null;
+            String parseError = null;
+            try {
+                ABIType ABIType = this.methodMap.get(i).returns.parsedType;
+                decoded = ABIType.decode(abiEncoded);
+            } catch (Exception e) {
+                parseError = e.getMessage();
+            }
+            retList.add(new ReturnValue(
+                    this.transactionList.get(i).txn.txID(),
+                    abiEncoded,
+                    decoded,
+                    parseError
+            ));
+        }
+
+        return new ExecuteResult(txInfo.confirmedRound, this.getTxIDs(), retList);
+    }
+
+    private static boolean checkLogRet(byte[] logLine) {
+        if (logLine.length < 4)
+            return false;
+        for (int i = 0; i < 4; i++) {
+            if (logLine[i] != ABI_RET_HASH[i])
+                return false;
+        }
+        return true;
+    }
+
+    public static class ExecuteResult {
+        public Long confirmedRound;
+        public List<String> txIDs;
+        public List<ReturnValue> methodResults;
+
+        public ExecuteResult(Long confirmedRound, List<String> txIDs, List<ReturnValue> methodResults) {
+            this.confirmedRound = confirmedRound;
+            this.txIDs = txIDs;
+            this.methodResults = methodResults;
+        }
+    }
+
+    public static class ReturnValue {
+        public String txID;
+        public byte[] rawValue;
+        public Object value;
+        public String parseError;
+
+        public ReturnValue(String txID, byte[] rawValue, Object value, String parseError) {
+            this.txID = txID;
+            this.rawValue = rawValue;
+            this.value = value;
+            this.parseError = parseError;
+        }
+    }
+
+    public interface TxnSigner {
+        int hashCode();
```

Comment:
I think it would be more flexible if this type did not force subclasses to implement `hashCode` and `equals`. The default implementations of these methods should work fine for comparison by reference (which is fine for now), and subclasses can always choose to implement `hashCode` and `equals` if comparison by value is important to them.